[Nguyen Phu Hao] iP

VN-Hao opened the pull request:
### Qi - The chat bot

> “Life is too complicated not to be orderly.” – Martha Stewart ([source](https://dansilvestre.com/productivity-quotes/#10_%E2%80%9CLife_is_too_complicated_not_to_be_orderly%E2%80%9D_%E2%80%93_Martha_Stewart))

Qi can help manage your tasks in a more systematic way it can:

- classify tasks into: todo, deadline and event
- mark a specific task as done or undone
- find the task we need by key words

Start using Qi now by the following steps:

1. download the app from here
2. open **Command Prompt**/ **Windows PowerShell**
3. use `cd` command to navigate to the directory containg the app (can learn more from [here](https://www.digitalcitizen.life/command-prompt-how-use-basic-commands/))
4. done! :partying_face:

Features:

- [x] Managing tasks
- [ ] Chat UI (_comming soon_)
- [ ] Notification (_comming soon_)

Feel free to look through the source code from [here](https://github.com/VN-Hao/ip) and feedback to me!

Here is my current driver code:
```ruby
public static void main(String[] args) {
    new Qi("data/tasks.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

VN-Hao committed:
New chatbot with simple greetings

VN-Hao committed:
echo commands

VN-Hao committed:
Add, store and display tasks

VN-Hao committed:
Mark and unmark task, reorganize classes

VN-Hao committed:
Categorize tasks into todos, deadlines and events

VN-Hao committed:
Change input, EXPECTED and runtest file for UI Testing

VN-Hao committed:
Handle simple errors

VN-Hao committed:
Allow deletions of tasks

VN-Hao committed:
Minor changes in variable names. Rewrite some functions

VN-Hao committed:
Allow saving list to hard disk

VN-Hao committed:
Merge branch 'branch-Level-7'

VN-Hao committed:
Change type of deadline from String to java.time.LocalDate

VN-Hao committed:
Merge branch 'branch-Level-8'

VN-Hao committed:
rewrite code by OOP principles

VN-Hao committed:
handle input exception

VN-Hao committed:
Merge branch 'branch-A-MoreOOP'

VN-Hao committed:
group java classes into packages

VN-Hao committed:
Merge branch 'branch-A-Packages'

VN-Hao committed:
Merge branch 'master' into add-gradle-support

VN-Hao committed:
Merge branch 'add-gradle-support'

VN-Hao committed:
add JUnit tests

VN-Hao committed:
add find feature

VN-Hao committed:
Change the code to follow standard

VN-Hao committed:
Write documentation for classes and methods

VN-Hao committed:
Merge branch 'branch-A-JavaDoc'

VN-Hao committed:
Resolve conflicts

VN-Hao committed:
Write documentation for FindCommand class

VN-Hao committed:
Resolve conflicts caused by adding new method

jingjie88 commented on src/main/java/qi/parser/Parser.java:
```diff
@@ -0,0 +1,123 @@
+     * @return Command to be executed
+     * @throws QiException If the user input does not match the required format.
+     */
+    public static Command parse(String command) throws QiException {
```
I like the clean code here 👍

VN-Hao replied:
🙇

jingjie88 commented on src/main/java/qi/ui/Ui.java:
```diff
@@ -0,0 +1,141 @@
+/**
+ * Represents the interactions of the chatbot with users.
+ */
+public class Ui {
```
I like the consistent naming of the methods 👍

VN-Hao replied:
Thank you 🙇

jingjie88 reviewed (commented):
LGTM!

oeggy03 commented on src/main/java/qi/Qi.java:
```diff
@@ -0,0 +1,64 @@
+            } catch (QiException e) {
+                ui.showError(e.getMessage());
+            } finally {
+                ui.showLine();
```
Nice use of the ui class to print stuff, good OOP

oeggy03 commented on src/main/java/qi/command/AddCommand.java:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @param taskDescription String description of a Todo.
+     */
+    public AddCommand(String taskDescription) {
```
Maybe you can abstract each command parser into its own class for better oop

oeggy03 commented on src/main/java/qi/command/ExitCommand.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Represents the last interaction of the chatbot with users.
+ */
+public class ExitCommand extends Command {
```
Nice use of inheritance

oeggy03 commented on src/main/java/qi/parser/Parser.java:
```diff
@@ -0,0 +1,123 @@
+    public static Command parse(String command) throws QiException {
+        if (command.equals("bye")) {
+            return new ExitCommand();
+        }
```
I think you can also use case switch or a command map, as multiple if else statements may take up a lot of space

oeggy03 commented on src/main/java/qi/qiexception/QiException.java:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Represents the exceptions specific to Qi - the chatbot.
+ */
+public class QiException extends Exception {
```
Maybe you can also create more specific exceptions, e.g. QiWriteError etc.

oeggy03 reviewed (commented):
LGTM! I really like the clean code and good use of OOP

VN-Hao committed:
Add checkstyle

VN-Hao committed:
Merge branch 'branch-A-CheckStyle'

VN-Hao committed:
Implement basic GUI

VN-Hao committed:
Merge branch 'branch-Level-10'

VN-Hao committed:
Make Qi greet when the program starts

VN-Hao committed:
Make some changes in GUI

The avatars of User and Qi are in rectangle because they are
displayed using javafx.scene.image.ImageView.

Change the shape of the avatars to circle by replacing
javafx.scene.image.ImageView with javafx.scene.shape.Circle.

Doing so will make the UI more pleasant to users.

VN-Hao committed:
Use assert feature to document assumptions

The method `load` in qi/storage/Storage.java has many assumptions about
the format of the data file loaded in.

Use `assert` statements to ensure that every line of the data file
follows the desired format.

Doing so will verify that the method `update` works as intended.

VN-Hao committed:
Improve code quality of some methods

VN-Hao committed:
Merge pull request #2 from VN-Hao/branch-A-Assertions

Use assert feature to document assumptions

VN-Hao committed:
Merge branch 'master' into branch-A-CodeQuality

Resolve conflicts caused by unupdated code

VN-Hao committed:
Merge pull request #3 from VN-Hao/branch-A-CodeQuality

Improve code quality of some methods

VN-Hao committed:
Add help command

Because there are many commands, new users may be confused when
they use the app for the first time.

Add a help command to tell the purpose of the app and list out
all of the current commands as well as their formats.

Doing so will help new users feel more convinient whenever they
forget the syntax of any command.

VN-Hao committed:
Merge pull request #4 from VN-Hao/branch-BCD-Extension

Add help command

VN-Hao committed:
Improve overall GUI

The avatar of the user and the texts do not change accordingly when
the app's window resizes.

Set the height and the width of the dialogBox to refer to the
scrollPane's dimensions.

This will make the GUI look more satisfying and flexible.

VN-Hao committed:
Write more JavaDoc and fix a bug caused by load

VN-Hao committed:
Handle errors caused by invalid task ID

VN-Hao committed:
Merge branch 'branch-A-MoreErrorHandling'

VN-Hao committed:
Change the title of the Github page

VN-Hao committed:
Write readme file into a user guide
